### PR TITLE
Feature/ratings

### PR DESCRIPTION
**Pull request by sirn-se**
Add optional review grading as part of comment.
Enabled by setting Site.settings.useReview to true.

### Server

The only restriction is that a Review must be linked to a Comment. Review grade can be any integer.

### Client

The client adds a couple of other restrictions;
* Only one Review per Site/User combination. Subsequent gradings overwrites existing.
* Review grade is limited to 1-5 (or no grading at all)
* Possibly, the limitations above can later be implemented as settings rather then hadrwired into the client.

### CODE REVIEW

**Comment by sirn-se** [comment]
What result does coverage/coveralls compare with?! The 2 "missing lines" in models/user.js wasn't introduced here.

File: migrations/20151020151002_reviews.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+'use strict'
+
+exports.up = function(knex) {
+  return knex.schema.createTable('reviews', function(table) {
+    table.increments()
+    table.integer('grade').notNullable()
+
+    table.integer('userId')
+    table.integer('pageId')
+    table.integer('commentId')
```
Comment (fiddur) [inline migrations/20151020151002_reviews.js]: Add foreign key references.

File: models/review.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    this.grade = grade
+    return this.$query().patch({grade: grade, commentId: commentId})
+  }
+
+  Review.prototype.getPage = async(function() {
+    if (this.page) {return this.page}
+
+    await(this.$loadRelated('page'))
+    return await(this.$loadRelated('page')).page
+  })
```
Comment (fiddur) [inline models/review.js]: Double await on same loadRelated…

File: public/client.js [hunk 3]
```diff
@@ -126,6 +126,36 @@
...
+
+        gradeItem.addEventListener('click', function(cl) {
+          reviewGrade = this.value
+          for (var i = 1; i <= 5; i++) {
+            reviewGradeLabelItems[i].className = i <= reviewGrade ? 'active' : ''
+          }
+        })
+      }
+      commentDiv.appendChild(reviewDiv)
+    }
```
Comment (fiddur) [inline public/client.js]: I'd say this is too big to inline; make a small function to return reviewDiv :)

File: models/review.js (hunk 2 above)
Comment (sirn-se) [inline models/review.js]: The entire function is copied as-is from model/comment.js, but you´re right. The non-returning await(…) seems superfluous in both cases.

File: public/client.js [hunk 4]
```diff
@@ -384,6 +448,69 @@
...
+            return review
+          }, function(error) {
+            if (error instanceof ForbiddenError) {
+              // Lets offer login and retry
+              return User.offerLogin(site.server, error.call)
+                .then(function (reviewJson) {
+                  var review = JSON.parse(reviewJson)
+                  return review
+                })
+            }
```
Comment (fiddur) [inline public/client.js]: It really would be prudent to make a wrapper for this retrial soon, it's copied to too many places now!

File: public/client.js [hunk 5]
```diff
@@ -475,6 +615,29 @@
...
+    var userSpan = document.createElement('span')
+    userSpan.className = 'commenter_name'
+    userSpan.appendChild(document.createTextNode(comment.user.displayName || ''))
+    return userSpan;
+  }
+
+  function reviewDiv(review) {
+    var reviewDiv = document.createElement('div')
+    reviewDiv.className = 'review'
+    for (var i = 1; i <= 5; i++) {
```
Comment (fiddur) [inline public/client.js]: This `5` is in several places.  Get a single source of truth ™ :)

File: public/client.js [hunk 6]
```diff
@@ -475,6 +615,29 @@
...
+      reviewDiv.appendChild(reviewGradeSpan(review.grade >= i))
+    }
+    return reviewDiv;
+  }
+
+  function reviewGradeSpan(active) {
+    var gradeSpan = document.createElement('span')
+    gradeSpan.innerHTML = '★'
```
Comment (fiddur) [inline public/client.js]: This `★` is in several places.  Get a single source of truth ™ :)

File: routes/reviews.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+ * GNU-AGPL-3.0
+ */
+
+'use strict'
+
+var async = require('asyncawait/async')
+var await = require('asyncawait/await')
+
+//// TODO remove Q
+var Q          = require('q')
```
Comment (fiddur) [inline routes/reviews.js]: Do you use Q in this file?

File: routes/reviews.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+var Handlebars = require('handlebars')
```
Comment (fiddur) [inline routes/reviews.js]: Handlebars is not used in this file.

File: routes/reviews.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+var FS         = require('fs')
+var path       = require('path')
+var markdown   = require('markdown').markdown
+
+module.exports = function (app, model, config) {
+
+  // Get by comment
+  app.get('/sites/:site/pages/:page/reviews/', async(function(req, res) {
```
Comment (fiddur) [inline routes/reviews.js]: Get by comment?

File: public/client.js (hunk 4 above)
Comment (sirn-se) [inline public/client.js]: The client could do well with some re-factoring, it´s somewhat messy at this point. Preferably it should be split by logic and rendering, enabling different renderers as plug-ins (or equivalent).
But that´s not within the scope of this feature, can´t do everything at once :)

**Comment by fiddur** [comment]
Looks good.  I haven't tried it out myself, but I'll merge it to master now without making a version  tag.

**Pull request merged**